File: OC/ATCountingView.m

//
//  ATCountView.m
//  Test
//
//  Created by ApesTalk on 2019/1/13.
//  Copyright © 2019年 https://github.com/ApesTalk All rights reserved.
//

#import "ATCountingView.h"

@interface UIView (ATFrameExtension)
@property (nonatomic, assign) CGFloat at_x;
@property (nonatomic, assign) CGFloat at_y;
@property (nonatomic, assign) CGFloat at_w;
@property (nonatomic, assign) CGFloat at_h;
@property (nonatomic, assign) CGSize at_size;
@property (nonatomic, assign) CGPoint at_origin;
@end

@implementation UIView(ATFrameExtension)
- (void)setAt_x:(CGFloat)at_x
{
    CGRect frame = self.frame;
    frame.origin.x = at_x;
    self.frame = frame;
}

- (CGFloat)at_x
{
    return self.frame.origin.x;
}

- (void)setAt_y:(CGFloat)at_y
{
    CGRect frame = self.frame;
    frame.origin.y = at_y;
    self.frame = frame;
}

- (CGFloat)at_y
{
    return self.frame.origin.y;
}

- (void)setAt_w:(CGFloat)at_w
{
    CGRect frame = self.frame;
    frame.size.width = at_w;
    self.frame = frame;
}

- (CGFloat)at_w
{
    return self.frame.size.width;
}

- (void)setAt_h:(CGFloat)at_h
{
    CGRect frame = self.frame;
    frame.size.height = at_h;
    self.frame = frame;
}

- (CGFloat)at_h
{
    return self.frame.size.height;
}

- (void)setAt_size:(CGSize)at_size
{
    CGRect frame = self.frame;
    frame.size = at_size;
    self.frame = frame;
}

- (CGSize)at_size
{
    return self.frame.size;
}

- (void)setAt_origin:(CGPoint)at_origin
{
    CGRect frame = self.frame;
    frame.origin = at_origin;
    self.frame = frame;
}

- (CGPoint)at_origin
{
    return self.frame.origin;
}
@end

@interface ATCountingView () <CAAnimationDelegate>
@property (nonatomic, strong) UILabel *frontLabel;///< current show label
@property (nonatomic, strong) UILabel *backLabel;///< hide or animate label
@end

@implementation ATCountingView
- (instancetype)initWithFrame:(CGRect)frame
{
    if(self = [super initWithFrame:frame]){
        self.clipsToBounds = YES;
        _backLabel = [[UILabel alloc]initWithFrame:CGRectMake(0, 0, frame.size.width, frame.size.height)];
        _backLabel.textAlignment = NSTextAlignmentCenter;
        _backLabel.text = @"0";
        [self addSubview:_backLabel];
        
        _frontLabel = [[UILabel alloc]initWithFrame:CGRectMake(0, 0, frame.size.width, frame.size.height)];
        _frontLabel.textAlignment = NSTextAlignmentCenter;
        _frontLabel.text = @"0";
        [self addSubview:_frontLabel];
    }
    return self;
}

- (void)layoutSubviews{
    [super layoutSubviews];
    _frontLabel.at_size = self.at_size;
    _backLabel.at_size = self.at_size;
}

- (void)setTextColor:(UIColor *)textColor{
    _textColor = textColor;
    _frontLabel.textColor = textColor;
    _backLabel.textColor = textColor;
    
}

- (void)setFont:(UIFont *)font{
    _font = font;
    _frontLabel.font = font;
    _backLabel.font = font;
}

- (void)setTextAlignment:(NSTextAlignment)textAlignment
{
    _textAlignment = textAlignment;
    _frontLabel.textAlignment = textAlignment;
    _backLabel.textAlignment = textAlignment;
}

- (void)setNumber:(NSInteger)number{
    _number = MIN(9, MAX(0, number));
    _frontLabel.text = [NSString stringWithFormat:@"%ld", _number];
    _backLabel.text = [NSString stringWithFormat:@"%ld", _number];
}

- (void)animateFromNumber:(NSInteger)fromNumber toNumber:(NSInteger)toNumber perDuration:(NSTimeInterval)perDuration{
    //考虑到cycle模式下有可能大于9也有可能小于0
    NSInteger f = fromNumber;
    if(f >= 10){
        f = f-10;
    }else if (f < 0) {
        f = f+10;
    }
    _frontLabel.text = [NSString stringWithFormat:@"%ld", f];
    _frontLabel.at_y = 0;
    if(fromNumber < toNumber){
        //increase
        //考虑到cycle模式下有可能大于9
        NSInteger value = fromNumber+1;
        if(value >= 10){
            value = value-10;
        }
        _backLabel.text = [NSString stringWithFormat:@"%ld", value];
        _backLabel.at_y = self.at_h;
        _backLabel.hidden = NO;
        [UIView animateWithDuration:perDuration animations:^{
            _frontLabel.transform = CGAffineTransformTranslate(_frontLabel.transform, 0, -self.at_h);
            _backLabel.transform = CGAffineTransformTranslate(_backLabel.transform, 0, -self.at_h);
        } completion:^(BOOL finished) {
            if(!finished){
                return;
            }
            
            UILabel *tmpLabel = self.frontLabel;
            self.frontLabel = self.backLabel;
            self.backLabel = tmpLabel;
            self.backLabel.hidden = YES;
            _number++;
            if(_number < toNumber){
                [self animateFromNumber:_number toNumber:toNumber perDuration:perDuration];
            }else{
                _number = value;
                if([_delegate respondsToSelector:@selector(animationDidStopForCountView:)]){
                    [_delegate animationDidStopForCountView:self];
                }
            }
        }];
    }else if (fromNumber > toNumber){
        //decrease
        //考虑到cycle模式下有可能小于0
        NSInteger value = fromNumber-1;
        if(value < 0){
            value = value+10;
        }
        _backLabel.text = [NSString stringWithFormat:@"%ld", value];
        _backLabel.at_y = -self.at_h;
        _backLabel.hidden = NO;
        [UIView animateWithDuration:perDuration animations:^{
            _frontLabel.transform = CGAffineTransformTranslate(_frontLabel.transform, 0, self.at_h);
            _backLabel.transform = CGAffineTransformTranslate(_backLabel.transform, 0, self.at_h);
        } completion:^(BOOL finished) {
            if(!finished){
                return;
            }
            
            UILabel *tmpLabel = self.frontLabel;
            self.frontLabel = self.backLabel;
            self.backLabel = tmpLabel;
            self.backLabel.hidden = YES;
            _number--;
            if(_number > toNumber){
                [self animateFromNumber:_number toNumber:toNumber perDuration:perDuration];
            }else{
                _number = value;
                if([_delegate respondsToSelector:@selector(animationDidStopForCountView:)]){
                    [_delegate animationDidStopForCountView:self];
                }
            }
        }];
    }
}

- (void)animateToNumber:(NSUInteger)number duration:(NSTimeInterval)duration{
    [_frontLabel.layer removeAllAnimations];
    [_backLabel.layer removeAllAnimations];

    if([_delegate respondsToSelector:@selector(animationDidStartForCountView:)]){
        [_delegate animationDidStartForCountView:self];
    }
    if (number == _number){
        //no animtion
        if([_delegate respondsToSelector:@selector(animationDidStopForCountView:)]){
            [_delegate animationDidStopForCountView:self];
        }
    }else{
        //判断_number执行加法和减法哪种方式离number更近
        NSInteger realToNumber = number;
        if(self.isCycle){
            if(number > _number){
                NSInteger addTimes = number - _number;
                NSInteger reduceTimes = _number + 10 - number;
                if(addTimes <= reduceTimes){
                    realToNumber = number;
                }else{
                    realToNumber = number - 10;
                }
            }else{
                NSInteger reduceTimes = _number - number;
                NSInteger addTimes = 10 - _number + number;
                if(reduceTimes <= addTimes){
                    realToNumber = number;
                }else{
                    realToNumber = number + 10;
                }
            }
        }
        [self animateFromNumber:_number toNumber:realToNumber perDuration:fabs(duration/(realToNumber-_number))];
    }
}

@end

File: ATCountingView/ATCountingView/ATCountingView.swift
//
//  ATCountingView.swift
//  ATCountingView
//
//  Created by ApesTalk on 2019/1/13.
//  Copyright © 2019年 https://github.com/ApesTalk All rights reserved.
//

import UIKit

protocol ATCountViewProtocol: class {
    func animationDidStartForCountView(view:ATCountingView)
    func animationDidStopForCountView(view:ATCountingView)
}

class ATCountingView: UIView {
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.clipsToBounds = true
        configUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    var textColor = UIColor.black
    var font = UIFont.systemFont(ofSize: 17)
    var textAlignment = NSTextAlignment.center
    var number: Int = 0
    var cycle = false // 是否循环，默认false。循环时会自动判断执行加法还是减法距离目标数字近，比如2->9会执行减法动画
    weak var delegate: ATCountViewProtocol?
    
    private var frontLabel = UILabel()
    private var backLabel = UILabel()
    
    func configUI() {
        frontLabel.textAlignment = NSTextAlignment.center
        frontLabel.text = "0"
        self.addSubview(frontLabel)
        
        backLabel.textAlignment = NSTextAlignment.center
        backLabel.text = "0"
        self.addSubview(backLabel)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        frontLabel.setAt_size(at_size: self.at_size())
        backLabel.setAt_size(at_size: self.at_size())
    }
    
    func setTextColor(color: UIColor) {
        textColor = color
        frontLabel.textColor = textColor
        backLabel.textColor = textColor
    }
    
    func setFont(f: UIFont) {
        font = f
        frontLabel.font = font
        backLabel.font = font
    }
    
    func setTextAlignment(alignment: NSTextAlignment) {
        textAlignment = alignment
        frontLabel.textAlignment = textAlignment
        backLabel.textAlignment = textAlignment
    }
    
    func setNumber(num: Int) {
        number = min(9, max(0, num))
        frontLabel.text = String(num)
        backLabel.text = String(num)
    }
    
    private func animateFromNumber(fromNumber: Int, toNumber: Int, perDuration: TimeInterval){
        //考虑到cycle模式下有可能大于9也有可能小于0
        var f = fromNumber
        if f >= 10{
            f = f-10
        }else if f < 0 {
            f = f+10
        }
        frontLabel.text = String(f)
        frontLabel.setAt_y(at_y: 0)
        if fromNumber < toNumber {
            //increase
            //考虑到cycle模式下有可能大于9
            var value = fromNumber+1
            if value >= 10{
                value = value-10
            }
            backLabel.text = String(value)
            backLabel.setAt_y(at_y: self.at_h())
            backLabel.alpha = 1
            UIView.animate(withDuration: perDuration, animations: {
                self.frontLabel.transform = self.frontLabel.transform.translatedBy(x: 0, y: -self.at_h())
                self.backLabel.transform = self.backLabel.transform.translatedBy(x: 0, y: -self.at_h())
            }) { (finished) in
                let tmpLabel = self.frontLabel
                self.frontLabel = self.backLabel
                self.backLabel = tmpLabel
                self.backLabel.alpha = 0
                self.number = self.number+1
                if self.number < toNumber {
                    self.animateFromNumber(fromNumber: self.number, toNumber: toNumber, perDuration: perDuration)
                }else{
                    self.number = value
                    self.delegate?.animationDidStopForCountView(view: self)
                }
            }
        }else{
            //decrease
            //考虑到cycle模式下有可能小于0
            var value = fromNumber-1
            if value < 0 {
                value = value+10
            }
            backLabel.text = String(value)
            backLabel.setAt_y(at_y: -self.at_h())
            backLabel.alpha = 1
            UIView.animate(withDuration: perDuration, animations: {
                self.frontLabel.transform = self.frontLabel.transform.translatedBy(x: 0, y: self.at_h())
                self.backLabel.transform = self.backLabel.transform.translatedBy(x: 0, y: self.at_h())
            }) { (finished) in
                let tmpLabel = self.frontLabel
                self.frontLabel = self.backLabel
                self.backLabel = tmpLabel
                self.backLabel.alpha = 0
                self.number = self.number-1
                if self.number > toNumber {
                    self.animateFromNumber(fromNumber: self.number, toNumber: toNumber, perDuration: perDuration)
                }else{
                    self.number = value
                    self.delegate?.animationDidStopForCountView(view: self)
                }
            }
        }
    }
    
    func animateToNumber(toNumber: Int, duration: TimeInterval) {
        frontLabel.layer.removeAllAnimations()
        backLabel.layer.removeAllAnimations()
        self.delegate?.animationDidStartForCountView(view: self)
   
        if (toNumber == self.number){
            //no animtion
            self.delegate?.animationDidStopForCountView(view: self)
        }else{
            //判断_number执行加法和减法哪种方式离number更近
            var realToNumber = toNumber
            if self.cycle {
                if toNumber > self.number {
                    let addTimes = toNumber - self.number
                    let reduceTimes = self.number + 10 - toNumber
                    if addTimes <= reduceTimes {
                        realToNumber = toNumber
                    }else{
                        realToNumber = toNumber - 10
                    }
                }else{
                    let reduceTimes = self.number - toNumber
                    let addTimes = 10 - self.number + toNumber
                    if reduceTimes <= addTimes {
                        realToNumber = toNumber
                    }else{
                        realToNumber = toNumber + 10
                    }
                }
            }
            let count: Double = Double(realToNumber-self.number)
            let pertimes: Double = Double(fabs(duration/count))
            self.animateFromNumber(fromNumber: self.number, toNumber: realToNumber, perDuration: pertimes)
        }
    }

}
